### PR TITLE
Move Storage and StorageImpl to c10

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14022 Move Scalar and ScalarType to c10/core&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13015236/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14023 Fix include paths for Scalar.h and ScalarType.h&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081609/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14024 Move ScalarTypeUtils.h to c10&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081604/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14058 Move UniqueVoidPtr to c10&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081602/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14059 Move Allocator.h to c10&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081606/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14060 Fix include paths for Allocator.h&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081605/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#14061 Move Storage and StorageImpl to c10**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081608/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14062 Fix include paths for Storage.h and StorageImpl.h&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081603/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14327 Move TensorTypeId to c10/core&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13131338/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14070 Fix include paths for TensorTypeId.h and TensorTypeIdRegistration.h&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081610/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14141 Remove fake dependencies from TensorImpl to caffe2&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13113129/)

Differential Revision: [D13081608](https://our.internmc.facebook.com/intern/diff/D13081608/)

### CODE REVIEW

File: c10/core/Scalar.h [hunk 1]
```diff
@@ -86,7 +87,7 @@ class C10_API Scalar {
     //    error: designator may not specify a non-POD subobject
     double z[2];
   } v;
-  friend struct Type;
+  friend struct at::Type;
```
Comment (ezyang) [inline c10/core/Scalar.h]: at::

Comment (ezyang) [reply c10/core/Scalar.h]: I'm pretty sure we can just un-friend this...

File: c10/core/Storage.cpp [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#include <c10/core/Storage.h>
```
Comment (ezyang) [inline c10/core/Storage.cpp]: Kill this file

File: c10/core/StorageImpl.h [hunk 3]
```diff
@@ -0,0 +1,227 @@
+#pragma once
+
+#include <c10/core/Allocator.h>
+#include <c10/core/ScalarType.h>
+#include <c10/core/ScalarTypeUtils.h>
+
+#include <c10/util/intrusive_ptr.h>
+
+namespace at {
+struct Type;
```
Comment (ezyang) [inline c10/core/StorageImpl.h]: You should have gotten rid of the Type forward declaration *before* this patch ;) Now you get to hear me complain about it.

**Review by ezyang — APPROVED** [review]
accepted but with comments

**Comment by smessmer** [comment]
Comments addressed in https://github.com/pytorch/pytorch/pull/14432 and https://github.com/pytorch/pytorch/pull/14139

**Pull request closed without merging**